10-query/01b-pattern-negation.md

```diff
@@ -0,0 +1,334 @@
+
+```
+traversable sub entity,
+    has index;
+    plays from,
+    plays to;
+
+node sub traversable;
+
+link sub relation, relates from, relates to;
+```
+
+Then we can define two nodes as being reachable if there exists a link between them:
+
+```
+reachable sub relation, relates from, relates to;
+reachabilityA sube rule,
+when {
+    (from: $x, to: $y) isa link;
+},
+then {
+    (from: $x, to: $y) isa reachable;
+};
+
+reachabilityB sub rule,
+when {
+    (from: $x, to: $z) isa link;
+    (from: $z, to: $y) isa reachable;
+},
+then {
+    (from: $x, to: $y) isa reachable;
+};
+```
+
+Consequently, with the use of negation we can define links that are indirect:
+
+```
+indirect-link sub relation, relates from, relates to;
+indirect-link-rule sub rule,
+when {
+    (from: $x, to: $y) isa reachable;
+    not {(from: $x, to: $y) isa link;};
+},
+then {
+    (from: $x, to: $y) isa indirect-link;
+};
+```
+
+or, mark the unreachable nodes by specifying the following rule:
+
+```
+unreachable sub relation, relates from, relates to;
+unreachability-rule
+when {
+    $x isa node;
+    $y isa node;
+    not {(from: $x, to: $y) isa reachable;};
+},
+then {
+    (from: $x, to: $y) isa unreachable;
+};
+```
+
+Please note the explicit addition of the `node` types in the body of the rule. This is to to ensure the boundedness condition
+is satisfied as well as to maintain the expected meaning of `unreachable` relation.
+
+
+
+## Monotonicity of reasoning with negation
+
+In this subsection we will address an important property of reasoning being the monotonicity property.
+We say that reasoning is monotonic if previously derived facts remain true upon addition of new knowledge. Up to this point Graql reasoning was monotonic. However with
+the addition of pattern negation this property no longer holds. We can illustrate this fact in a popular example:
+
+Let's start with our knowledge to be the following:
+
+- birds fly unless they are abnormal
+- penguins are abnormal birds
+- penguins are birds
+- Tweety is a bird
+
+This can be summarised as:
+
+```
+flies(X) :-  bird(X), ¬abnormal(X)
+abnormal(X) :- penguin(X)
+bird(X) :- penguin(X)
+```
+
+We can capture this information in Graql terms:
+
+Defining our little schema first:
+
+```
+bird sub entity;
+penguin sub bird;
+flies sub entity;
+abnormal isa entity;
+```
+
+Accompanying it with relevant rules:
+
+```
+flying-rule sub rule,
+when{
+    $x isa bird;
+    not {$x isa abnormal;};
+},
+then{
+    $x isa flies;
+}
+
+abnormal-rule sub rule,
+when{
+    $x isa penguin;
+},
+then{
+    $x isa abnormal;
+}
+```
+
+In consequence of the above knowledge, we can establish that Tweety flies, i. e. if we query:
+
+```
+match $x isa flies;get;
+```
+the Tweety concept will be returned.
+
+Let us now look what will happen if we add an extra bit of information - we will specialise Tweety to be a penguin. If we now repeat our flying query, we will
```

Comment:
Can we also do this with code and repeat the same query as before for total clarity. Then, make a punchier point, something along these lines to labour the point:
_We see that when adding an additional fact, we lose a fact that was previously derived -- that `Tweety isa flies`._

Reply:
Should we add code for this?